test/test_torch.py

```diff
@@ -5863,6 +5863,25 @@ def test_batch_norm_cpu_inference(self):
             self.assertEqual(output3, output1)
             self.assertEqual(output3, output2)
 
+    def test_tensor_grad_warnings(self):
+        dummy = torch.empty(1)
+
+        with warnings.catch_warnings(record=True) as w:
+            # Accessing .grad on leaf
+            dummy.requires_grad_()
+            foo = dummy.grad
+            self.assertEqual(len(w), 0)
+
+            # Accessing .grad on non-leaf
+            dummy = dummy.clone()
+            foo = dummy.grad
+            self.assertEqual(len(w), 1)
+
+            # Accessing .grad on non-leaf that retain gradients
+            dummy.retain_grad()
+            foo = dummy.grad
+            self.assertEqual(len(w), 1)
```

Comment:
nit: Since you are just catching any warnings, there could be a situtation where an unrelated warning is raised but not the relevant one?

Reply:
Since we only call `retain_grad()` and `.grad` in valid cases. We actually don't want other warnings to be raised here. Otherwise, we have an overly verbose warning.